{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Working with data packages (DP/dp)\n",
    "## Introduction\n",
    "A `dp` is a `.json` file containg the data contextual information: `location` (local path) or `URI` (remote), the `schema` (dataset structure) and its `metadata` (data about the dataset). The `dp` allows us to organise out data in a compact way, which is consisted with the [FAIR](https://www.nature.com/articles/sdata201618) principles of data stewardship and use. It promote data openness and research reproducibility. There are two broad approaches for dealing with `dp`:\n",
    "- Web-based graphical app\n",
    "- Programming interfaces (different languages)\n",
    "\n",
    "For this case will look at the `Python` interface for working with `dp`. You may look at [datapackage-py](https://github.com/frictionlessdata/datapackage-py) GitHub repository for detailed instructions.\n",
    "\n",
    "## Creating a data package\n",
    "The Python `datapackage` is available for this purpose and can be installed using the pip Python installation manager using the command `pip install datapackage` from the Linux/MAC terminal. I presime you have Python already installed in your computer.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'profile': 'tabular-data-package',\n",
       " 'resources': [{'name': 'hrma_optimisation',\n",
       "   'path': 'https://raw.githubusercontent.com/ousodaniel/data/master/Spacies_table.csv',\n",
       "   'profile': 'tabular-data-resource',\n",
       "   'schema': {'fields': [{'name': 'Common name',\n",
       "      'type': 'string',\n",
       "      'format': 'default',\n",
       "      'title': 'c_name',\n",
       "      'description': 'Common name'},\n",
       "     {'name': 'Species',\n",
       "      'type': 'string',\n",
       "      'format': 'default',\n",
       "      'title': 'species',\n",
       "      'description': 'Scientific name'},\n",
       "     {'name': 'Family',\n",
       "      'type': 'string',\n",
       "      'format': 'default',\n",
       "      'title': 'family',\n",
       "      'description': 'Taxonomic family'},\n",
       "     {'name': 'Tissue',\n",
       "      'type': 'integer',\n",
       "      'format': 'default',\n",
       "      'title': 'tissue_sample',\n",
       "      'description': 'Solid sample counts'},\n",
       "     {'name': 'Blood',\n",
       "      'type': 'integer',\n",
       "      'format': 'default',\n",
       "      'title': 'blood_sample',\n",
       "      'description': 'Whole blood sample counts'},\n",
       "     {'name': 'Total',\n",
       "      'type': 'integer',\n",
       "      'format': 'default',\n",
       "      'title': 'tot_sample',\n",
       "      'description': 'Blood plus tissue sample counts'},\n",
       "     {'name': 'Seq_id_COI',\n",
       "      'type': 'string',\n",
       "      'format': 'default',\n",
       "      'title': 'gi_genbank_coi',\n",
       "      'description': 'Unique COI seq identifier for sequences deposited in GenBank'},\n",
       "     {'name': 'Seq_id_cyt b',\n",
       "      'type': 'string',\n",
       "      'format': 'default',\n",
       "      'title': 'gi_genbank_cytb',\n",
       "      'description': 'Unique cyt b seq identifier for sequences deposited in GenBank'}],\n",
       "    'missingValues': ['']},\n",
       "   'title': 'Supplementary Table 1',\n",
       "   'description': 'Samples used in the HRM reference optimisations',\n",
       "   'format': 'CSV',\n",
       "   'encoding': 'UTF-8'},\n",
       "  {'name': 'hrma_validation',\n",
       "   'path': 'https://raw.githubusercontent.com/ousodaniel/data/master/368445%202Jan2020_rev_ms.csv',\n",
       "   'profile': 'tabular-data-resource',\n",
       "   'schema': {'fields': [{'name': 'S/ID',\n",
       "      'type': 'string',\n",
       "      'format': 'default',\n",
       "      'title': 's_id',\n",
       "      'description': 'Sample lab id/name'},\n",
       "     {'name': 'COI',\n",
       "      'type': 'string',\n",
       "      'format': 'default',\n",
       "      'title': 'coi',\n",
       "      'description': 'COI High-resolution melting analysis prediction of species identity of sample'},\n",
       "     {'name': 'cyt b',\n",
       "      'type': 'string',\n",
       "      'format': 'default',\n",
       "      'title': 'cyt_b',\n",
       "      'description': 'cyt b High-resolution melting analysis prediction of species identity of sample'},\n",
       "     {'name': '16S rRNA ',\n",
       "      'type': 'string',\n",
       "      'format': 'default',\n",
       "      'title': '16_rrna',\n",
       "      'description': '16S rRNA High-resolution melting analysis prediction of species identity of sample'},\n",
       "     {'name': 'Predicted by HRMA',\n",
       "      'type': 'string',\n",
       "      'format': 'default',\n",
       "      'title': 'hrma_predict',\n",
       "      'description': 'Sample id by combined analysis prediction'},\n",
       "     {'name': 'Vouchered ID',\n",
       "      'type': 'string',\n",
       "      'format': 'default',\n",
       "      'title': 'confirmed_id',\n",
       "      'description': 'Vouchered sample id by sequencing'}],\n",
       "    'missingValues': ['']},\n",
       "   'description': 'Blind identification of DNA extracts',\n",
       "   'title': 'Supplementary Table 2',\n",
       "   'format': 'CSV',\n",
       "   'encoding': 'UTF-8'}],\n",
       " 'keywords': ['coi', 'cyt b', '16S rRNA', 'hrma', 'bushmeat'],\n",
       " 'name': 'hrma_validation',\n",
       " 'title': 'HRMA Validation',\n",
       " 'description': 'Tittle: Three-gene PCR and high-resolution melting analysis for differentiating vertebrate species mitochondrial DNA for biodiversity research and complementing forensic surveillance\\n\\nAbstract: Reliable molecular identification of vertebrate species from morphologically unidentifiable tissue is critical for the prosecution of illegally-traded wildlife products, conservation-based biodiversity research, and identification of blood-meal hosts of hematophagous invertebrates. However, forensic identification of vertebrate tissue relies on sequencing of the mitochondrial cytochrome oxidase I (COI) ‘barcode’ gene, which remains costly for purposes of screening large numbers of unknown samples during routine surveillance. Here, we adapted a rapid, low-cost approach to differentiate 10 domestic and 24 wildlife species that are common in the East African illegal wildlife products trade based on their unique high-resolution melting profiles from COI, cytochrome b, and 16S ribosomal RNA gene PCR products. Using the approach, we identified (i) giraffe among covertly sampled meat from Kenyan butcheries, and (ii) forest elephant mitochondrial sequences among savannah elephant reference samples. This approach is being adopted for high-throughput pre-screening of potential bushmeat samples in East African forensic science pipelines.\\n\\n* Previously published sequence',\n",
       " 'homepage': 'https://raw.githubusercontent.com/ousodaniel/data/master/368445%202Jan2020%20JV_DO_7-01-20.csv?token=AGSN3RTHAQ6PTNZXFJYQLKC6FEPPS',\n",
       " 'contributors': [{'title': 'Ouso', 'role': 'author'}],\n",
       " 'licenses': [{'name': 'CC-BY-SA-4.0',\n",
       "   'title': 'Creative Commons Attribution Share-Alike 4.0',\n",
       "   'path': 'https://creativecommons.org/licenses/by-sa/4.0/'}]}"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from datapackage import Package\n",
    "# Inferring dp from tabular data\n",
    "package = Package()# Init an empty dp\n",
    "package2 = Package()\n",
    "package.infer('**/*.csv')# Infers a dp from all csv datasets\n",
    "package2.infer('**/SurgicalDataFrame.csv')\n",
    "package.descriptor# Displays the dp\n",
    "#Save dps\n",
    "package.save('data_dp.json')\n",
    "package2.save('SurgicalDataFrame_dp.json')\n",
    "# Creating a dp from scratch\n",
    "package3 = Package()# Init an empty dp\n",
    "# Use the descriptor attribute to assign attributes of the dp\n",
    "package3.descriptor['name'] = 'period-table'\n",
    "package3.descriptor['title'] = 'Periodic Table'\n",
    "package3.descriptor['resources'] = 'data.csv'\n",
    "package3.descriptor# Displays the dp\n",
    "package3.save('arbitrary_dp.json')\n",
    "\n",
    "package.descriptor['resources'][0]# Get first resource\n",
    "\n",
    "package.descriptor['resources'][1]# Get second resource\n",
    "# Loading an existing dp\n",
    "package_ld = Package('datapackage-hrma_validation.json')\n",
    "package_ld.descriptor\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Validating tabular data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from goodtables import goodtables"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
